tensorflow/core/ops/collective_ops.cc

```diff
@@ -142,6 +142,35 @@ REGISTER_OP("CollectiveReduceV2")
     .SetIsDistributedCommunication()
     .SetShapeFn(shape_inference::UnchangedShape);
 
+REGISTER_OP("CollectiveReduceScatterV2")
+    .Input("input: T")
+    .Output("data: T")
+    .Attr("T: {bfloat16, float, float16, float64, int32, int64}")
+    .Input("group_size: int32")
+    .Input("group_key: int32")
+    .Input("instance_key: int32")
+    .Input("ordering_token: Nordering_token * resource")
+    .Attr("merge_op: {'Min', 'Max', 'Mul', 'Add'}")
+    .Attr("final_op: {'Id', 'Div'}")
+    .Attr("communication_hint: string = 'auto'")
+    .Attr("timeout_seconds: float = 0")
+    .Attr("Nordering_token: int >= 0 = 0")
+    .Attr("max_subdivs_per_device: int = -1")
+    .SetIsStateful()
+    .SetIsDistributedCommunication()
+    .SetShapeFn([](shape_inference::InferenceContext* c) {
+      // Scalar input is not supported.
+      shape_inference::ShapeHandle unused;
+      TF_RETURN_IF_ERROR(c->WithRankAtLeast(c->input(0), 1, &unused));
+      // This output should have the same shape as its input except the first
+      // dimension is unknown, since the group size is unknown.
+      shape_inference::ShapeHandle out;
+      TF_RETURN_IF_ERROR(
+          c->ReplaceDim(c->input(0), /*dim_index=*/0, c->UnknownDim(), &out));
```

Comment:
I am a bit surprised that ReduceScatter doesn't need an input for the local output shape, which can be handy if we are to support uneven sharding down the road. Does the NCCL backend allow such an option?

Reply:
It doesn't seem like `ncclReduceScatter` supports uneven sharding https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/colls.html#c.ncclReduceScatter

Reply:
Sure. There isn't much we can do about this then.

Reply:
I found that in horovod, they implemented uneven sharding using a group of ncclReduce ops instead of ncclReduceScatter:
https://github.com/horovod/horovod/blob/master/horovod/common/ops/nccl_operations.cc#L1298

Do you think it's worth adding an output_shape input for this future use case? It looks like XlaReduceScatter doesn't support uneven sharding so perhaps we don't need this?

Reply:
I agree with your assessment! Let's skip outputshape for now. We can always add it if we have to.